### PR TITLE
Add explicit Prepare step to motivate randomization

### LINKED ISSUE #148: Randomization appears without context or rationale
[Section 5](https://cfrg.github.io/draft-irtf-cfrg-blind-signatures/draft-irtf-cfrg-rsa-blind-signatures.html#section-5) says *what* and *how* in ways that seem to be pretty solid, but doesn't contextualize the mechanism at all.  [Section 6](https://cfrg.github.io/draft-irtf-cfrg-blind-signatures/draft-irtf-cfrg-rsa-blind-signatures.html#section-6) doubles down on this and defines message formats that include randomization.

However, there is also nothing in preceding sections that might indicate why you might want to do this.

It is my understanding that the definition of message content is application-specific, which might interact poorly with any randomization scheme that is applied without considering those interactions.  One does not simply randomize messages without understanding what effect that randomization might have on the consumer of the message.

The definitions in [Section 6](https://cfrg.github.io/draft-irtf-cfrg-blind-signatures/draft-irtf-cfrg-rsa-blind-signatures.html#section-6) appear to run counter to the goal of using an off-the-shelf RSASSA-PSS verifier, because they require that the consumer understand that randomization has occurred.

Furthermore, the notes on use don't really scope the applicability of the mechanisms.  I am unable to use the text in this document to determine why randomization or not is necessary because I don't know where this might be applied.  One application is cited, but that seems to be just an example: what system properties are necessary for the overall design to be applicable?  And, what subset of those cases demand the use of randomization?

The goal here appears to be one that is motivated by Privacy Pass use cases, where a single message might need to be used multiple times but also be indistinguishable from each other.  It might be that there are no cases where having a signing oracle that doesn't see what it signs makes sense outside of this sort of anonymous ticket vending protocols, but the draft appears to assume this setting without directly saying so.

Consequently, without leaning heavily on an assumption that this is for Privacy Pass, these two sections were very hard to process.

### PR DESCRIPTION
**Pull request by chris-wood**
Closes #148

This adds an explicit `Prepare` step to the protocol, which is either a no-op (in the case of no randomization applied) or it augments the message with randomness. Rather than making the protocol description a complicated series of "if your application needs this, then first invoke the randomizer routine...." and so on, it seemed easier to just wrap up this step in a generic "preparation" function. The named variants then just say what the prepare step does, i.e., it either spits out the input message or one that has been augmented with randomness.

@martinthomson, @smhendrickson: does this seem more approachable to you?

### CODE REVIEW

**Review by martinthomson — COMMENTED** [review]
This helps by tying the randomization into the main flow properly.

I am also interested in being clearer about why a client might insist on randomization.

In the "message-entropy" section where you say:

> Based on this fact, using the core protocol functions in {{core-protocol}} is possibly unsafe, unless one of the following conditions are met:

I might instead say:

> A client that uses this protocol [note: there is only a core protocol now] might be vulnerable to attack from a malicious signer and verifier unless it is able to ensure that either:
> 1. The signer has proven that its public key is honestly generated. [...the rest of this point]
> 2. The input message has a value that the signer is unable to guess.  That is, the client has added a high-entropy component that was not available to the signer prior to them choosing their signing key.

For the second point, it is my understanding that it is not sufficient for the value to be high entropy.  Rather it has to be difficult for the signer to guess the value that a client might use.

The consequences of that choice for consumers of signed messages is also worth exploring.  Something like:

> The randomization of messages protects clients from malicious signers.  A verifier that accepts randomized messages needs to remove the random component from the signed part of messages before processing.

Of course, a verifier is very much limited in terms of what it might do with information that the signer probably never saw, so most uses of the message content are limited.  An application might choose to have clients include a fixed string to further prevent signatures from being applied outside of the context that the client chooses, which in turn might be a subset of what the signer enables through its signature.

**Comment by chris-wood** [comment]
Thanks again @martinthomson. These are good suggestions. I've accepted most of them, but didn't add further text on what information the verifier sees since, as you sort of hint, the verifier sees everything at the end and blindness means it can't really do much with that by coordinating with the signer.

**Comment by martinthomson** [comment]
I wasn't concerned as much about what the verifier does or does not see, but about what it might need to do to get to the signed "message".  Randomization effectively adds junk to the message that the verifier needs to know about and strip out.

**Comment by chris-wood** [comment]
@jedisct1 and @FredericJacobs, can you please review?

**Comment by chris-wood** [comment]
> I wasn't concerned as much about what the verifier does or does not see, but about what it might need to do to get to the signed "message". Randomization effectively adds junk to the message that the verifier needs to know about and strip out.

Got it. I added text that hopefully makes this expectation for verifiers more clear.

File: draft-irtf-cfrg-rsa-blind-signatures.md [hunk 1]
```diff
@@ -381,69 +416,46 @@ Steps:
...
+part of messages before processing.
 
 Applications that require deterministic signatures can use the RSABSSA-SHA384-PSSZERO-Deterministic
 variant, but only if their input messages have high entropy. Applications that use
 RSABSSA-SHA384-PSSZERO-Deterministic SHOULD carefully analyze the security implications,
 taking into account the possibility of adversarially generated signer keys as described in
 {{message-entropy}}. When it is not clear whether an application requires deterministic or
-randomized signatures, applications SHOULD use one of the recommended variants with message
-randomization.
+randomized signatures, applications SHOULD use one of the recommended variants with randomized message preparation.
```
Comment (martinthomson) [inline draft-irtf-cfrg-rsa-blind-signatures.md:458]:
```suggestion
randomized signatures, applications SHOULD use one of the variants with randomized message preparation.
```

**Review by martinthomson — APPROVED** [review]

**Review by jedisct1 — APPROVED** [review]

**Comment by FredericJacobs** [comment]
I like this phrasing. LGTM

**Pull request merged**